pep-0692.rst

```diff
@@ -0,0 +1,344 @@
+
+The ``kwargs`` parameter inside the function should be treated exactly the same way as any other ``TypedDict``.
+
+Function calls with standard dictionaries
+-----------------------------------------
+
+Calling a function that has ``**kwargs`` typed using the ``**kwargs: **Movie`` sytnax with a standard dictionary must raise an error. For example:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+    movie: dict[str, str] = {"name": "Life of Brian", "year": 1979}
+    foo(**movie)  # WRONG! Movie is of type dict[str, str]
+
+    typedMovie: Movie = {"name": "The Meaning of Life", "year": 1983}
+    foo(**typedMovie)  # OK!
+
+
+However, if the dictionary is completely untyped it is up to the type checker whether to raise an error or not:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+    movie = {"name": "Life of Brian", "year": 1979} 
+
+    foo(**movie)  # Error/Success dependant on the type checker.
+
+
+Keyword collisions
+------------------
+
+A ``TypedDict`` that is used to type ``**kwargs`` could potentially contain keys that are already defined in the function's signature. If the duplicate name is a standard argument, an error should be raised. If the duplicate name is a positional only argument, no errors should be raised. For example:
+
+::
+
+    def foo(name, **kwargs: **Movie) -> None: ...     # WRONG! "name" will always bind to the first parameter.
+
+    def foo(name, /, **kwargs: **Movie) -> None: ...  # OK! "name" is a positional argument, so **kwargs can contain a "name" keyword
+
+
+Required and non-required keys
+------------------------------
+
+By default all keys in a ``TypedDict`` are required. This behaviour can be overriden by setting the dictionary's ``total`` parameter as ``False``:
+
+::
+
+    class Movie(TypedDict, total=False):
+        title: str
+        year: int 
+
+    dictionary: Movie = {"title": "Life of Brian"}         # OK! "year" is not required.
+    dictionary = {"title": "Life of Brian", "year": 1979}  # OK!
+
+Moreover, PEP 655 introduced new type qualifiers - ``typing.Required`` and ``typing.NotRequired`` - that enable specyfing whether a particular key is required or not:
+
+::
+
+    class Movie(TypedDict):
+        title: str
+        year: NotRequired[int]
+
+When using a ``TypedDict`` to type ``**kwargs`` all of the required and non-required keys should correspond to required and non-required function keyword parameters. Therefore, if a required key is not supported by the caller, then an error must be raised.
+
+Assignment
+----------
+A function typed using the ``**kwargs: **Movie`` construct can be assigned to another callable type only if they are compatible. This can happen in the following cases:
+
+1. Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination ``TypedDict`` is assignable to the source ``TypedDict`` and the rest of the parameters are the compatible.
```

Comment:
```suggestion
1. Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination ``TypedDict`` is assignable to the source ``TypedDict`` and the rest of the parameters are compatible.
```

Reply:
Also, I think this gets the variance wrong. Consider two callables `f1: (int) -> bool` and `f2: (object) -> bool`. `f1` is *not* assignable to `f2`, because when you have an `f2`, you can pass it any object, but `f1` must have an int. On the other hand, `f2` is assignable to `f1`. Functions parameters behave contravariantly.

For analogous reasons, for two functions `(**TD1) -> ...` and `(**TD2) -> ...`, the first function is assignable to the second if `TD2` is assignable to `TD1`, not the other way around.

(2) also gets the variance wrong.

Last, you ignore the return type, but it also affects assignability. However, it is covariant.

Reply:
I think the way I worded it was confusing.

> For analogous reasons, for two functions `(**TD1) -> ...` and `(**TD2) -> ...`, the first function is assignable to the second if `TD2` is assignable to `TD1`, not the other way around.

The first function is assignable to the second, so the first function is the "source" and the second one is the "destination". So the destination `TypedDict` (`TD2`) has to be assignable to the source TypedDict (`TD1`) - I think this is the source (pun not intended) of the error - what I meant is:

`Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination function's ``TypedDict`` is assignable to the source function's ``TypedDict`` and the rest of the parameters are compatible.`

Let me try to find a better way to phrase it.